File: src/regression.py
import numpy as np
from fastcore.utils import *


class Regression:

    def __init__(self, n_iter, lr, regularization=None):
        store_attr()

    def _init_weight(self, n_features):
        w = np.random.normal(loc=0, scale=1, size=n_features)
        return w

    def fit(self, x, y):
        x = (x - np.mean(x, axis=0)) / (np.std(x, axis=0))
        x = np.c_[np.ones(x.shape[0]), x]
        self.w = self._init_weight(x.shape[1])
        for _ in range(self.n_iter):
            y_pred = x @ self.w
            if self.regularization is not None:
                mse = np.mean(0.5 * (y - y_pred) ** 2 + self.reguralization(self.w))
            else:
                mse = np.mean(0.5 * (y - y_pred) ** 2)
                grad_w = -(y - y_pred) @ x
                self.w -= self.lr * grad_w

    def predict(self, x):
        x = np.c_[np.ones(x.shape[0]), x]
        y_pred = x @ self.w
        return y_pred


class LassoRegularization:
    def __init__(self, alpha):
        store_attr()

    def __call__(self, w):
        return self.alpha * np.linalg.norm(w)

    def grad(self, w):
        return self.alpha * np.sign(w)


class RidgeRegularization:

    def __init__(self, alpha):
        store_attr()

    def __call__(self, w):
        return 0.5 * self.alpha * w @ w

    def grad(self, w):
        return self.alpha * w


class ElasticRegression:
    def __init__(self, alpha, contribution=0.5):
        store_attr()

    def __call__(self, w):
        lasso = self.contribution * np.linalg.norm(w)
        ridge = (1 - self.contribution) * w @ w
        return self.alpha * (lasso + ridge)

    def grad(self, w):
        lasso = self.contribution * np.sign(w)
        ridge = (1 - self.contribution) * w
        return self.alpha * (lasso + ridge)

File: julia/regression.jl
using Statistics, Distributions


function init_weight(n_features::Int)
    w = rand(Normal(0, 1), n_features)
    return w
end


function regression(x, y, n_iter, λ)
    x = (x .- mean(x, dims=1))./ (std(x, dims=1))
    x = hcat(ones(size(x)[1]), x)
    w = init_weight(size(x)[2])
    error = Any[]
    for i=1:n_iter
        y_pred = x * w
        mse = mean(0.5 * (y - y_pred).^2)
        push!(error, mse)
        grad_w = - transpose(x) * (y - y_pred)
        print(size(grad_w))
        w .-= λ .* grad_w
        println(size(w))
    end
    return error, w 
end


x = rand(Normal(3, 11), (100, 3))
y = rand(Normal(2,5), 100)

regression(x, y, 100, 0.02)
